system/trace/trace_dump.c

```diff
@@ -0,0 +1,570 @@
+
+  tctx = g_dump_task_ctx;
+  g_dump_task_ctx = NULL;
+  while (tctx != NULL)
+    {
+      ntctx = tctx->next;
+      free(tctx);
+      tctx = ntctx;
+    }
+}
+
+/****************************************************************************
+ * Name: copy_task_name
+ ****************************************************************************/
+
+#if CONFIG_TASK_NAME_SIZE > 0
+static void copy_task_name(FAR char *dst, FAR const char *src)
+{
+  char c;
+  int i;
+
+  /* Replace space to underline
+   * Text trace data format cannot use a space as a task name.
+   */
+
+  for (i = 0; i < CONFIG_TASK_NAME_SIZE; i++)
+    {
+      c = *src++;
+      if (c == '\0')
+        {
+          break;
+        }
+
+      *dst++ = (c == ' ') ? '_' : c;
+    }
+
+  *dst = '\0';
+}
+#endif
+
+/****************************************************************************
+ * Name: get_task_context
+ ****************************************************************************/
+
+FAR static struct trace_dump_task_context_s *get_task_context(pid_t pid)
+{
+  FAR struct trace_dump_task_context_s **tctxp;
+  tctxp = &g_dump_task_ctx;
+  while (*tctxp != NULL)
+    {
+      if ((*tctxp)->pid == pid)
+        {
+          return *tctxp;
+        }
+
+      tctxp = &((*tctxp)->next);
+    }
+
+  /* Create new trace dump task context */
+
+  *tctxp = (FAR struct trace_dump_task_context_s *)
+           malloc(sizeof(struct trace_dump_task_context_s));
+  if (*tctxp != NULL)
+    {
+      (*tctxp)->next = NULL;
+      (*tctxp)->pid = pid;
+      (*tctxp)->syscall_nest = 0;
+      (*tctxp)->name[0] = '\0';
+    }
+
+  return *tctxp;
+}
+
+/****************************************************************************
+ * Name: get_task_name
+ ****************************************************************************/
+
+static const char *get_task_name(pid_t pid)
+{
+  FAR struct trace_dump_task_context_s *tctxp;
+
+  tctxp = get_task_context(pid);
+  if (tctxp != NULL && tctxp->name[0] != '\0')
+    {
+      return tctxp->name;
+    }
+
+  return "<noname>";
+}
+
+/****************************************************************************
+ * Name: trace_dump_header
+ ****************************************************************************/
+
+static void trace_dump_header(FILE *out,
+                              FAR struct note_common_s *note)
+{
+  pid_t pid;
+  uint32_t systime = note->nc_systime[0] +
+                     (note->nc_systime[1] << 8) +
+                     (note->nc_systime[2] << 16) +
+                     (note->nc_systime[3] << 24);
+#ifdef CONFIG_SMP
+  int cpu = note->nc_cpu;
+#else
+  int cpu = 0;
+#endif
+
+  pid = g_dump_cpu_ctx[cpu].current_pid;
+
+  fprintf(out, "%8s-%-3u [%d] %3u.%09u: ",
+          get_task_name(pid), get_pid(pid), cpu,
+          systime / (1000 * 1000 / CONFIG_USEC_PER_TICK),
+          (systime % (1000 * 1000 / CONFIG_USEC_PER_TICK))
+            * CONFIG_USEC_PER_TICK * 1000);
```

Comment:
why need multiple by CONFIG_USEC_PER_TICK?

Reply:
`systime % (1000 * 1000 / CONFIG_USEC_PER_TICK)` gets the number of ticks less than a second.
Because it is a unit of ticks, we need to multiply `CONFIG_USEC_PER_TICK * 1000` to get a unit of nanosecs which the ftrace format requires.